Added Snippets

wxtim opened the pull request:
Closes #20 

## Cylc Snippet Generator Script

Added script (`bin/get_keywords`) which parses `cylc/flow/workflow/cfgspec` into a `cylc-snippets.json` file for use by vscode:

![Screenshot from 2024-06-17 13-38-50](https://github.com/cylc/vscode-cylc/assets/26465611/f29d4334-e169-41a7-aafb-8850b8154196)


## Generation of updates to the Cylc Snippets

Added a workflow to automate running the script on demand or once a month, to ensure that the data is up to date:

Example PR https://github.com/wxtim/vscode-cylc/pull/3


**Check List**

- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

wxtim committed:
added some handy snippets.

wxtim committed:
simplify

MetRonnie commented on .github/workflows/update-snippets.yml:
```diff
@@ -0,0 +1,70 @@
+  # Ideally people making changes should re-run this task, but probably
+  # won't remember - I think once a month is probably often enough.
+  schedule:
+    - cron: "* * 15 * *"
```
https://crontab.guru/#*_*_15_*_*

> “At every minute on day-of-month 15.”

```suggestion
    - cron: "51 03 15 * *" # 03:51 on the 15th of the month
```

MetRonnie commented on .github/workflows/update-snippets.yml:
```diff
@@ -0,0 +1,70 @@
+          echo "create_pr=false" >> $GITHUB_OUTPUT
+        fi
+
+  commit:
+    runs-on: ubuntu-latest
+    needs: [snippets]
+    if: needs.snippets.outputs.create_pr == 'true'
+    steps:
+
+    - uses: actions/checkout@v4
+
+    - name: Install Cylc
+      run: pip install cylc-flow
+
+    - name: Update Snippets
+      run: ./bin/get_keywords > cylc-snippets.json
+
+    - name: Configure git
+      uses: cylc/release-actions/configure-git@v1
```
Duplication - why not just one job?

wxtim replied:
I want the second job only to run if the first one finishes - probably needs to be an artifact storing workflow or something - any pointers?

MetRonnie replied:
But you can do that in one job
```yml
- name: Checkout PR branch etc
  if: steps.set_output.output.create_pr == 'true'
  ...
```

wxtim committed:
Update .github/workflows/update-snippets.yml

Co-authored-by: Ronnie Dutta <61982285+MetRonnie@users.noreply.github.com>

wxtim committed:
f

MetRonnie commented on cylc.language-configuration.json:
```diff
@@ -27,5 +27,92 @@
+		],
+		"description": ["Pre formed task section"]
+	},
+	"spice": {
+		"prefix": ["spice"],
+		"body": [
+			"    [[${1:root}]]",
+			"        script = \"\"\"",
+			"            ${2:true}",
+			"        \"\"\"",
+			"        platform = ${3:spice}",
+			"        execution time limit = ${4:PT1M}",
+			"        execution retry delays = ${5:PT1M}",
+			"        [[[directives]]]",
+			"            --mem=${6:256M}",
+			"            --ntasks=${7:2}",
+		],
+		"description": ["Pre formed task section for a spice task"]
+	},
```
Met Office specific stuff shouldn't be in the extension. Instead we could document internally how users can add this snippet themselves.

wxtim replied:
This whole section was a mistake - I meant only to commit the auto generated thing...

wxtim committed:
remove stuff committed in error

MetRonnie commented on bin/get_keywords:
```diff
@@ -0,0 +1,28 @@
+    print(
+        'FAIL: This script requires an environment with cylc.flow installed'
+        '\n    ``pip install cylc-flow``'
+    )
```
`sys.exit(1)`?

wxtim replied:
```suggestion
    )
    sys.exit(1)
```

MetRonnie replied:
`import sys`?

MetRonnie commented on .github/workflows/update-snippets.yml:
```diff
@@ -0,0 +1,56 @@
+      run: |
+        if [[ -n $(git status cylc-snippets.json) ]]; then
+          echo "create_pr=true" >> $GITHUB_OUTPUT
+        else
+          echo "create_pr=false" >> $GITHUB_OUTPUT
```
Don't actually need this
```suggestion
```

MetRonnie commented on .github/workflows/update-snippets.yml:
```diff
@@ -0,0 +1,56 @@
+      if: steps.set_output.outputs.create_pr == 'true'
+      run : |
+        git commit -a -m "Update Cylc Config Snippets" -m "Workflow: ${{ github.workflow }}, run: ${{ github.run_number }}"
+        git push origin update_snippets -f
```
I'm not sure it should force push? 

Btw, no need to specify the remote and branch because you have used `cylc/release-actions/configure-git`
```suggestion
        git push
```

MetRonnie reviewed (commented):
When I run the extension with F5, I can't get any snippets to work?

wxtim committed:
Apply suggestions from code review

Co-authored-by: Ronnie Dutta <61982285+MetRonnie@users.noreply.github.com>

wxtim commented:
> When I run the extension with F5, I can't get any snippets to work?

Not sure - works for me. Any thoughts on how we debug it further?

MetRonnie commented:
Does it have to be registered like here? https://github.com/microsoft/vscode-extension-samples/blob/5651637c527e07173bd066f5fb7e171ef4616cab/completions-sample/src/extension.ts#L51

Note you should probably clear your user snippets before testing, if you have added the same snippets to your `~/.Code` or wherever they go

wxtim commented:
> Note you should probably clear your user snippets before testing, if you have added the same snippets to your `~/.Code` or wherever they go

Yes - that's why it was working for me.

> Does it have to be registered like here? https://github.com/microsoft/vscode-extension-samples/blob/5651637c527e07173bd066f5fb7e171ef4616cab/completions-sample/src/extension.ts#L51

I've been working off https://github.com/microsoft/vscode-extension-samples/tree/main/snippet-sample.

On further investigation I've found that the problem was that because my personal snippets were working I tried to make the snippets.json generator more minimal and wrecked the snippets file which didn't actually work.

Give it another go.

wxtim committed:
make the extension actually work

MetRonnie commented on cylc-snippets.json:
```diff
@@ -0,0 +1,672 @@
+        "body": "[[[directives]]]",
+        "description": "Job runner (batch scheduler) directives.\n\nThe default value is set in the global config: :cylc:conf:`global.cylc[platforms][<platform name>][directives]`.\n\nSupported for use with job runners:\n\n- pbs\n- slurm\n- loadleveler\n- lsf\n- sge\n- slurm_packjob\n- moab\n\nDirectives are written to the top of the job script in the correct format\nfor the job runner.\n\nSpecif ...
+    },
+    "<directive>": {
+        "prefix": "<directive>",
+        "body": "<directive> = ${1:*value unset*}",
+        "description": "Example directives for the built-in job runner handlers are shown in\n:ref:`AvailableMethods`."
+    },
```
Ah, should probably strip the "many" settings by detecting items that begin and end with `<`/`>`

MetRonnie reviewed (commented):
Nice, snippets working now

wxtim committed:
avoid namespace items in config

MetRonnie commented on cylc-snippets.json:
```diff
@@ -24,14 +24,14 @@
-        "prefix": "<custom metadata>",
-        "body": "<custom metadata> = ${1:}",
-        "description": "Any user-defined metadata item.\n\nThese, like title, description and URL. can be passed to\ntask event handlers to be interpreted according to your\nneeds. For example, the value of an \"importance\" item could\ndetermine how an event handler responds to task failure\nevents."
+    "__MANY__": {
+        "prefix": "__MANY__",
+        "body": "__MANY__ = ${1:*value unset*}",
+        "description": ".. deprecated:: 7.8.7/7.9.2\n\n   Parameter environment templates have moved to\n   :cylc:conf:`flow.cylc[runtime]\n   [<namespace>][environment]`."
```
Hm, this is another one we don't want

wxtim committed:
allow different hints with the same name

MetRonnie committed:
Improvements to snippets script

- Script writes snippets directly
- Make GH Actions workflow more robust
- Add VSCode task for building snippets

MetRonnie committed:
Rename snippets build script

MetRonnie committed:
Consolidate VSCode build tasks

MetRonnie commented:
Got a PR against this one at https://github.com/wxtim/vscode-cylc/pull/5

MetRonnie committed:
Update README

wxtim committed:
Merge pull request #5 from MetRonnie/snippets

Snippets improvements

MetRonnie reviewed (approved):
Awesome, I am personally going to find this really handy 🥳
